File: second.rb
def find_final_destination(routes)
    return "обратитесь к специалисту" if routes.empty?
  
    route_map = {}
    is_destination = {}
  
    routes.each do |from, to|
      if from == to
        return "обратитесь к специалисту"
      end
  
      route_map[from] = to
      is_destination[to] = true
    end
  
    start_point = nil
    routes.each do |from, to|
      unless is_destination.key?(from)
        start_point = from
        break
      end
    end
  
    return "обратитесь к специалисту" if start_point.nil?
  
    current = start_point
    longest_path_end = current
  
    while route_map.key?(current)
      current = route_map[current]
      longest_path_end = current
    end
  
    longest_path_end
  end
  
  puts "Введите количество пар маршрутов: "
  n = gets.chomp.to_i
  
  routes = []
  n.times do |i|
    puts "Введите начальный пункт #{i + 1}: "
    from = gets.chomp
    puts "Введите конечный пункт #{i + 1}: "
    to = gets.chomp
    routes << [from, to]
  end
  
  result = find_final_destination(routes)
  puts "Итоговый конечный пункт самой длинной цепочки: #{result}"

File: second.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>

using namespace std;

string findFinalDestination(const vector<pair<string, string>>& routes) {
    if (routes.empty()) {
        return "обратитесь к специалисту";
    }

    unordered_map<string, string> routeMap;
    unordered_map<string, bool> isDestination;

    for (const auto& route : routes) {
        const string& from = route.first;
        const string& to = route.second;

        if (from == to) {
            return "обратитесь к специалисту";
        }

        routeMap[from] = to;
        isDestination[to] = true;
    }

    string startPoint;
    for (const auto& route : routes) {
        if (isDestination.find(route.first) == isDestination.end()) {
            startPoint = route.first;
            break;
        }
    }

    if (startPoint.empty()) {
        return "обратитесь к специалисту";
    }

    string current = startPoint;
    string longestPathEnd = current;

    while (routeMap.find(current) != routeMap.end()) {
        current = routeMap[current];
        longestPathEnd = current;
    }

    return longestPathEnd;
}

int main() {
    vector<pair<string, string>> routes;
    int n;

    cout << "Введите количество пар маршрутов: ";
    cin >> n;
    cin.ignore();

    for (int i = 0; i < n; ++i) {
        string from, to;
        cout << "Введите начальный пункт " << i + 1 << ": ";
        getline(cin, from);
        cout << "Введите конечный пункт " << i + 1 << ": ";
        getline(cin, to);
        routes.emplace_back(from, to);
    }

    string result = findFinalDestination(routes);
    cout << "Итоговый конечный пункт самой длинной цепочки: " << result << endl;

    return 0;
}